storage/innobase/include/trx0purge.h

```diff
@@ -55,80 +55,28 @@ Run a purge batch.
 @return number of undo log pages handled in the batch */
 ulint trx_purge(ulint n_tasks, ulint history_size);
 
-/** Rollback segements from a given transaction with trx-no
-scheduled for purge. */
-class TrxUndoRsegs {
-private:
-	typedef std::vector<trx_rseg_t*, ut_allocator<trx_rseg_t*> >
-		trx_rsegs_t;
-public:
-	typedef trx_rsegs_t::iterator iterator;
-	typedef trx_rsegs_t::const_iterator const_iterator;
-
-	TrxUndoRsegs() = default;
-
-	/** Constructor */
-	TrxUndoRsegs(trx_rseg_t& rseg)
-		: trx_no(rseg.last_trx_no()), m_rsegs(1, &rseg) {}
-	/** Constructor */
-	TrxUndoRsegs(trx_id_t trx_no, trx_rseg_t& rseg)
-		: trx_no(trx_no), m_rsegs(1, &rseg) {}
-
-	bool operator!=(const TrxUndoRsegs& other) const
-	{ return trx_no != other.trx_no; }
-	bool empty() const { return m_rsegs.empty(); }
-	void erase(iterator& it) { m_rsegs.erase(it); }
-	iterator begin() { return(m_rsegs.begin()); }
-	iterator end() { return(m_rsegs.end()); }
-	const_iterator begin() const { return m_rsegs.begin(); }
-	const_iterator end() const { return m_rsegs.end(); }
-
-	/** Compare two TrxUndoRsegs based on trx_no.
-	@param elem1 first element to compare
-	@param elem2 second element to compare
-	@return true if elem1 > elem2 else false.*/
-	bool operator()(const TrxUndoRsegs& lhs, const TrxUndoRsegs& rhs)
-	{
-		return(lhs.trx_no > rhs.trx_no);
-	}
-
-	/** Copy of trx_rseg_t::last_trx_no() */
-	trx_id_t trx_no= 0;
-private:
-	/** Rollback segments of a transaction, scheduled for purge. */
-	trx_rsegs_t m_rsegs{};
+struct trx_rseg_cmp
+{
+  /** Compare two trx_rseg_t* based on trx_no.
+  @param lhs first pointer to trx_rseg_t to compare
+  @param rhs second pointer to trx_rseg_t to compare
+  @return true if elem1 > elem2 else false.*/
+  bool operator()(const trx_rseg_t *lhs, const trx_rseg_t *rhs)
+  {
+    /* We can do the comparision without trx_rseg_t latching, because rseg last
+    commit is always set before pushing rseg to purge queue. */
+    return lhs->last_trx_no() > rhs->last_trx_no();
+  }
 };
 
 typedef std::priority_queue<
-	TrxUndoRsegs,
-	std::vector<TrxUndoRsegs, ut_allocator<TrxUndoRsegs> >,
-	TrxUndoRsegs>	purge_pq_t;
-
-/** Chooses the rollback segment with the oldest committed transaction */
-struct TrxUndoRsegsIterator {
-	/** Constructor */
-	TrxUndoRsegsIterator();
-	/** Sets the next rseg to purge in purge_sys.
-	Executed in the purge coordinator thread.
-	@retval false when nothing is to be purged
-	@retval true  when purge_sys.rseg->latch was locked */
-	inline bool set_next();
-
-private:
-	// Disable copying
-	TrxUndoRsegsIterator(const TrxUndoRsegsIterator&);
-	TrxUndoRsegsIterator& operator=(const TrxUndoRsegsIterator&);
-
-	/** The current element to process */
-	TrxUndoRsegs			m_rsegs;
-	/** Track the current element in m_rsegs */
-	TrxUndoRsegs::const_iterator	m_iter;
-};
+    trx_rseg_t *, std::vector<trx_rseg_t *, ut_allocator<trx_rseg_t *>>,
+    trx_rseg_cmp>
+    purge_pq_t;
```

Comment:
These will point to `trx_sys.rseg_array`, right? What if we make the underlying array `byte[TRX_SYS_N_RSEGS]` and allocate it statically? If `std::priority_queue` needs to maintain the number of elements, we won’t be able to use `std::array` but we could implement something similar on our own. Here is some code from another C++ program that I wrote:
```c++
template <typename T, size_t N>
class vararray
{
public:
  constexpr auto begin() const { return m_arr; }
  constexpr auto begin() { return m_arr; }
  constexpr auto end() const { return begin () + size (); }
  constexpr auto end() { return begin () + size (); }
  constexpr auto operator[](size_t s) const { return m_arr[s]; }
  constexpr void push_back(const T& piece) { m_arr[m_size++]= piece; }
  constexpr size_t size() const { return m_size; }

private:
  T m_arr[N];
  size_t m_size= 0;
};
```
The `trx_rseg_cmp` would then be something like this:
```c++
struct trx_rseg_cmp
{
  /** Compare two trx_rseg_t* based on trx_no.
  @param lhs first pointer to trx_rseg_t to compare
  @param rhs second pointer to trx_rseg_t to compare
  @return whether lhs>rhs */
  bool operator()(const byte lhs, const byte rhs)
  {
    /* We can compare without trx_rseg_t::latch, because rseg last
    commit is always set before pushing rseg to purge queue. */
    return trx_sys.rseg_array[lhs].last_trx_no() >
      trx_sys.rseg_array[rhs].last_trx_no();
  }
};
```